### PR TITLE
 Add support for using the OpenTelemetry chloggen tool

### PR DESCRIPTION
**Pull request by jvoravong**
### Description

This PR brings several enhancements to the Makefile, introducing new utility functions, support for `chloggen`, and pre-commit checks.

#### Updated Makefile with New Actions
- `install-tools`: Installs required tools for CI/CD workflows using Homebrew and Go. Currently supports only MacOS and Linux.
- Changelog Management with `chloggen`: Added utility functions to create and manage changelog entries in `.chloggen/`. These entries will be used to automatically populate `CHANGELOG.md` during each release.
  - Pre-commit Check: Added a pre-commit hook that fails if a changelog entry is required. This is triggered when the Helm chart templates or rendered examples are updated.
- Other Utility Functions: Added another utility `make help`.

### Testing

The changes were manually tested.

### Documentation

- Updated `CONTRIBUTING.md` to reflect the new Makefile actions and the introduction of `chloggen`.

`make help` sample:

```
jvoravong@X splunk-otel-collector-chart % make help

Usage:
  make <target>

General
  help             Display Makefile help information for all actions

Initialization
  install-tools    Install tools (macOS/Linux)

Build
  repo-update      Update Helm repositories to latest
  dep-build        Build the Helm chart with latest dependencies from the current Helm repositories
  render           Render the Helm chart with the examples as input

Changelog
  chlog-available  Validate the chloggen tool is available
  chlog-new        Creates or updates a YAML file under .chloggen
  chlog-validate   Validates changelog requirements for pull requests
  chlog-preview    Provide a preview of the generated CHANGELOG.md file for a release
  chlog-update     Creates an update to CHANGELOG.md for a release entry from content in .chloggen
```

`make chlog-release VERSION=0.85.0` sample:

```
# Changelog

## [0.85.0] - 2023-09-18

### 💡 Enhancements 💡

This Splunk OpenTelemetry Collector for Kubernetes release adopts the [Splunk OpenTelemetry Collector v0.84.0](https://github.com/signalfx/splunk-otel-collector/releases/tag/v0.84.0).

- `chart`: Add support for OpenTelemetry CHANGELOG.md generator tool, see [chloggen](https://github.com/open-telemetry/opentelemetry-operator/tree/main/.chloggen) ([#923](https://github.com/signalfx/splunk-otel-collector-chart/pull/923))
- `agent`: Update Splunk Fluentd HEC docker image to v1.3.3 ([#924](https://github.com/signalfx/splunk-otel-collector-chart/pull/924))

## [0.84.0] - 2023-09-11

```

### CODE REVIEW

File: .github/workflows/lint-test.yaml [hunk 1]
```diff
@@ -36,12 +36,12 @@ jobs:
...
       - name: Run chart-testing (lint)
-        run: ct lint --config=ct.yaml
+        run: make lint
 
       - name: Create kind cluster
         uses: helm/kind-action@v1.8.0
         # Only build a kind cluster if there are chart changes to test.
-        if: steps.list-changed.outputs.changed == 'true'
-
+        if: steps.list-chaznged.outputs.changed == 'true'
```
[the remark this answers is not in the document]
Comment (jvoravong) [reply .github/workflows/lint-test.yaml]: Good catch. Typo. Fixed.

[the remark this answers is not in the document]
Comment (jvoravong) [reply .github/workflows/lint-test.yaml]: Currently, we are not prioritizing upstreaming this code.

**Comment by jvoravong** [comment]
Going to adopt changes from https://github.com/signalfx/splunk-otel-collector-chart/pull/917 before the next commit for this PR.

File: .github/workflows/lint-test.yaml [hunk 2]
```diff
@@ -36,12 +36,12 @@ jobs:
...
       - name: Run chart-testing (lint)
-        run: ct lint --config=ct.yaml
+        run: make lint
 
       - name: Create kind cluster
         uses: helm/kind-action@v1.8.0
         # Only build a kind cluster if there are chart changes to test.
         if: steps.list-changed.outputs.changed == 'true'
-
+        # TODO: make this test runnable from a makefile command
```
Comment (atoulme) [inline .github/workflows/lint-test.yaml:45]: this looks unrelated, please fork to another PR

File: .github/workflows/lint-test.yaml [hunk 3]
```diff
@@ -36,12 +36,12 @@ jobs:
           fi
 
       - name: Run chart-testing (lint)
-        run: ct lint --config=ct.yaml
+        run: make lint
```
Comment (atoulme) [inline .github/workflows/lint-test.yaml:39]: this looks unrelated, please fork to another PR

File: .github/workflows/validate-changelog.yaml [hunk 4]
```diff
@@ -11,12 +11,17 @@ jobs:
...
-          config: /lint/config/changelog.yml
-          args: ./CHANGELOG.md
+      - name: Setup Helm
+        uses: azure/setup-helm@v3.5
+
+      - name: Setup ct (chart testing)
+        uses: helm/chart-testing-action@v2.4.0
+
+      - name: Run Helm chart-testing (lint)
+        run: make lint
```
Comment (atoulme) [inline .github/workflows/validate-changelog.yaml:21]: this looks unrelated, please fork to another PR

File: Makefile [hunk 5]
```diff
@@ -1,25 +1,112 @@
...
+          /^[a-zA-Z_0-9-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } \
+          /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) }' \
+          $(MAKEFILE_LIST)
+
+##@ Initialization
+# Tasks for setting up the project environment
+
+.PHONY: init
+init: install-tools ## Initialize the environment
+	# TODO: Add more execution steps here
```
Comment (atoulme) [inline Makefile:32]: is that TODO needed?

File: Makefile [hunk 6]
```diff
@@ -1,25 +1,112 @@
...
 	if ! (helm repo list | grep -q open-telemetry) ; then \
-		helm repo add open-telemetry https://open-telemetry.github.io/opentelemetry-helm-charts ;\
+		helm repo add open-telemetry https://open-telemetry.github.io/opentelemetry-helm-charts || exit 1; \
 	fi ;\
 	if ! (helm repo list | grep -q jetstack) ; then \
-		helm repo add jetstack https://charts.jetstack.io ;\
+		helm repo add jetstack https://charts.jetstack.io || exit 1; \
 	fi ;\
-	helm repo update open-telemetry jetstack ;\
+	helm repo update open-telemetry jetstack || exit 1; \
```
Comment (atoulme) [inline Makefile:52]: these all should move to a separate PR

File: Makefile [hunk 7]
```diff
@@ -1,25 +1,112 @@
...
-  OK=true ;\
-  DIR=helm-charts/splunk-otel-collector ;\
-	if ! helm dependencies list $$DIR | grep open-telemetry | grep -q ok ; then OK=false ; fi ;\
-	if ! helm dependencies list $$DIR | grep jetstack | grep -q ok ; then OK=false ; fi ;\
-	if ! $$OK ; then helm dependencies build $$DIR ; fi ;\
+	DEP_OK=true ;\
+	DIR=helm-charts/splunk-otel-collector ;\
+	if ! helm dependencies list $$DIR | grep open-telemetry | grep -q ok ; then DEP_OK=false ; fi ;\
+	if ! helm dependencies list $$DIR | grep jetstack | grep -q ok ; then DEP_OK=false ; fi ;\
+	if [ "$$DEP_OK" = "false" ] ; then helm dependencies build $$DIR || exit 1; fi ;\
```
Comment (atoulme) [inline Makefile:62]: separate PR

File: Makefile [hunk 8]
```diff
@@ -1,25 +1,112 @@
...
+
+.PHONY: lint
+lint: ## Lint the Helm chart with ct
+	@echo "Linting Helm chart..."
+	ct lint --config=ct.yaml || exit 1
+
+.PHONY: pre-commit
+pre-commit: render ## Test the Helm chart with pre-commit
+	@echo "Checking the Helm chart with pre-commit..."
+	pre-commit run --all-files || exit 1
```
Comment (atoulme) [inline Makefile:80]: all separate PR

File: Makefile [hunk 9]
```diff
@@ -1,25 +1,112 @@
...
+
+##@ Changelog
+# Tasks related to changelog management
+
+.PHONY: chlog-available
+chlog-available: ## Validate the chloggen tool is available
+	@if [ -z "$(CHLOGGEN)" ]; then \
+		echo "Error: chloggen is not available. Please run 'make install-tools' to install it."; \
```
Comment (atoulme) [inline Makefile:88]: should that be a prereq of that task?

File: ci_scripts/install-tools.sh [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    installed_version=$($tool --version 2>/dev/null)  # Try to get the version
+    if [ -z "$installed_version" ]; then  # If version is empty
+      # Fallback to file modification time
+      installed_version="UNKNOWN (Last updated: $(stat -c %y "$tool_path" 2>/dev/null || stat -f "%Sm" "$tool_path"))"
+    fi
+  else
+    installed_version="Not Installed"
+  fi
+
+  read -p "$tool version $installed_version is installed. Would you like to upgrade? (y/n): " yn
```
Comment (atoulme) [inline ci_scripts/install-tools.sh:70]: `install_tools` is interactive? Is it necessary?

File: ci_scripts/install-tools.sh [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+      exit 1
+      ;;
+  esac
+}
+
+# Function to install or upgrade a Homebrew-based tool
+install_or_upgrade_brew() {
+  local tool=$1
+  local installed_version=$(brew list $tool --versions | awk '{print $2}')
+  local latest_version=$(brew info --json=v1 "$tool" | jq -r '.[0].versions.stable')
```
Comment (atoulme) [inline ci_scripts/install-tools.sh:36]: Will we be upgrading to the latest version available for helm at all times?

File: Makefile (hunk 8 above)
Comment (jvoravong) [reply Makefile:80]: Understood. I'll separate changes into separate PRs where possible. For an efficient integration of the chloggen tool in a DRY manner, it's probable that other PRs will have to be merged first.

File: Makefile (hunk 9 above)
Comment (jvoravong) [reply Makefile:88]: I'll see if we can make this happen. Was worried about if a single "make install-tools" install failed or if we are running on Windows.

File: ci_scripts/install-tools.sh (hunk 10 above)
Comment (jvoravong) [reply ci_scripts/install-tools.sh:70]: The version control for tools in this implementation, particularly helm and chloggen, are not as robust as in other projects like [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7db0833d735851da11277677e0a955d95055bf83/Makefile.Common#L40). Our current implementation takes the path of least resistance and might encounter version compatibility issues for developers. In the long run, we should consider containerizing these tools or enhancing the "make install-tools" for greater portability.

File: ci_scripts/install-tools.sh (hunk 11 above)
Comment (jvoravong) [reply ci_scripts/install-tools.sh:36]: While most tool dependencies in this project aren't pinned to specific versions, we've found that using the latest versions generally works without issues. However, older dependencies might not support newer features. Although this might only present challenges in rare instances, it's something we can address as needed in the future.

File: .github/workflows/lint-test.yaml (hunk 2 above)
Comment (jvoravong) [reply .github/workflows/lint-test.yaml]: done

File: .github/workflows/validate-changelog.yaml (hunk 4 above)
Comment (jvoravong) [reply .github/workflows/validate-changelog.yaml]: done

File: Makefile (hunk 7 above)
Comment (jvoravong) [reply Makefile:60]: We do need some of these updates to keep Makefile actions consistent over time. Separated out what I could. Done.

File: Makefile (hunk 8 above)
Comment (jvoravong) [reply Makefile]: Same as above. Removed what I could. Kept some for consistency. Done.

File: ci_scripts/install-tools.sh (hunk 10 above)
Comment (jvoravong) [reply ci_scripts/install-tools.sh]: Removed human input. Updated logging to reflect current state and actions executed. Done.

File: ci_scripts/install-tools.sh (hunk 11 above)
Comment (jvoravong) [reply ci_scripts/install-tools.sh:41]: At this time I would recommend installing latest. If we invest more into developer tool version control I think we can pin the version for the involved tools. Done.

File: Makefile (hunk 6 above)
Comment (jvoravong) [reply Makefile:50]: Removed what I could. Kept some for consistency. Done.

File: Makefile (hunk 9 above)
Comment (jvoravong) [reply Makefile:73]: Would recommend keeping this as is for now, can improve in the future.

**Comment by jvoravong** [comment]
Addressed PR comments and separated code out where possible. Executing final testing in GitHub workflows and then this should be ready to review.

**Comment by atoulme** [comment]
Did you have a separate PR for the code you separated?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -61,16 +61,45 @@ To send us a pull request, please:
    reformat all the code, it will be hard for us to focus on your change.
 3. Please follow the versioning instructions found in the [RELEASE.md](https://github.com/signalfx/splunk-otel-collector-chart/blob/main/RELEASE.md).
 4. Ensure local tests pass and add new tests related to the contribution.
-5. Commit to your fork using clear commit messages.
-6. Send us a pull request, answering any default questions in the pull request
+5. Add a CHANGLOG.md entry (make chlog-new) if the behavior of this chart is altered, see [Changelog Guidelines](https://signalfx/splunk-otel-collector-chart/blob/main/CONTRIBUTING.md#changelog-guidelines).
```
Comment (atoulme) [inline CONTRIBUTING.md:64]:
```suggestion
5. Add a CHANGELOG.md entry (make chlog-new) if the behavior of this chart is altered, see [Changelog Guidelines](https://signalfx/splunk-otel-collector-chart/blob/main/CONTRIBUTING.md#changelog-guidelines).
```

**Review by atoulme — APPROVED** [review]
LGTM, one nit

**Comment by jvoravong** [comment]
> Did you have a separate PR for the code you separated?

Haven't created it yet. Will do shortly.
